### DISCUSSION TITLE
Problem with object copy/clone

### DISCUSSION THREAD

**Discussion by southernangel** [fabricjs/fabric.js#7743] [Q&A] [answered]
I'm trying to implement the copy/clone functionality that's on the demo page but running into an error on the statement with the clone function.

        objToCopy.clone((cloned: any) => {
          this.fabricClipboard = cloned;
        })

I'm getting an error "this._removeDefaultValues is not a function" on the clone 

<img width="278" alt="error on copy" src="https://user-images.githubusercontent.com/26886652/155903117-10f5a6c9-05ec-476f-ad7d-12ff80540f90.png">
;

I've extended the fabric object prototype with some extra functions and prototypes as well.

This is an Angular 13 project, and I'm no expert at javascript, muddle through it at best.  Just wondering if someone can point out what I need to do to solve this error.

Thanks

**Comment by ShaMan123** [comment, marked as answer]
Docs are outdated if you are working on v5+
#3684 
```js
  objToCopy.clone().then((cloned: any) => {
      this.fabricClipboard = cloned;
    })
```

**Reply by southernangel** [reply]
Still working on 4.6.? as I've still got a problem with the toDataURL causing an error in version 5

Thanks anyway.  Its something to do with the weird way I've tried to use fabricJS in an angular framework.  I've created a fabric.Object prototype and I don't think I've put all the methods in.

**Reply by southernangel** [reply]
Actually I suspect it's something missing from this toObject override to add extra properties to the base object, probably a call to the super object.  As I mentioned, I'm an angular developer, so I'm pretty weak with the raw javascript

  this.Fabric.Object.prototype.toObject = ((toObject) => {
    return (propertiesToInclude: any) => {
        propertiesToInclude = (propertiesToInclude || []).concat(
          [ 'id', 'parentID', 'designType', 'emblemID', 'emblemCode', 'emblemType',
            'minimumWidth', 'minimumHeight', 'productID', 'productCode', 'designType',
            'colourID', '_controlsVisibility', 'fontID' ]
        );
        return toObject.apply(this, [propertiesToInclude]);
    };
  }) (this.Fabric.Object.prototype.toObject);

**Reply by southernangel** [reply]
Just realised that the "this" in the 
return toObject.apply(this, [propertiesToInclude]);
was actually the angular component, NOT the fabric instance that its supposed to be

I guess I'm trying to objectify fabric, but getting it all wrong.

The aim of the toObject prototype is to get those custom attributes as default.

**Reply by southernangel** [reply]
Answering my own question, I've figured it out with the help of this article

https://www.oodlestechnologies.com/blogs/extending-native-javascript-prototypes-in-angular/

I've obviously still got a lot to learn :-)

**Reply by ShaMan123** [reply]
angular is a js framework so extending/overriding prototypes should be the same.
This might help: http://fabricjs.com/fabric-intro-part-3#subclassing
The `this` concept requires trial and error to master. From your code I see this - using arrow functions `(() => { this. })` is a context less environment meaning `this` is shared from the context the arrow function is called from **as opposed** to `function () { this. }`. Here `this` is the caller of the function. In your case that is what you want because the fabric object is calling the method you want to override so don't use arrow functions for this.
